pennylane_lightning_gpu/src/simulator/MPIWorker.hpp

```diff
@@ -0,0 +1,330 @@
+        /* void* */ d_extraWorkspace,
+        /* size_t */ extraWorkspaceSize));
+
+    size_t transferWorkspaceSize =
+        size_t(1) << (numLocalQubits < 26 ? (numLocalQubits) : 26);
+
+    transferWorkspaceSize =
+        std::max(minTransferWorkspaceSize, transferWorkspaceSize);
+    PL_CUDA_IS_SUCCESS(cudaMalloc(&d_transferWorkspace, transferWorkspaceSize));
+    PL_CUSTATEVEC_IS_SUCCESS(custatevecSVSwapWorkerSetTransferWorkspace(
+        /* custatevecHandle_t */ handle,
+        /* custatevecSVSwapWorkerDescriptor_t */ svSegSwapWorker,
+        /* void* */ d_transferWorkspace,
+        /* size_t */ transferWorkspaceSize));
+
+    if (nP2PDeviceBits != 0) {
+        cudaIpcMemHandle_t ipcMemHandle;
+        PL_CUDA_IS_SUCCESS(cudaIpcGetMemHandle(&ipcMemHandle, sv));
+        std::vector<cudaIpcMemHandle_t> ipcMemHandles(mpi_manager.getSize());
+
+        mpi_manager.Allgather<cudaIpcMemHandle_t>(ipcMemHandle, ipcMemHandles,
+                                                  sizeof(ipcMemHandle));
+        cudaIpcEventHandle_t eventHandle;
+        PL_CUDA_IS_SUCCESS(cudaIpcGetEventHandle(&eventHandle, localEvent));
+        // distribute event handles
+        std::vector<cudaIpcEventHandle_t> ipcEventHandles(
+            mpi_manager.getSize());
+
+        mpi_manager.Allgather<cudaIpcEventHandle_t>(
+            eventHandle, ipcEventHandles, sizeof(eventHandle));
+        //  get remove device pointers and events
+        size_t nSubSVsP2P = 1 << nP2PDeviceBits;
+        size_t p2pSubSVIndexBegin =
+            (mpi_manager.getRank() / nSubSVsP2P) * nSubSVsP2P;
+        size_t p2pSubSVIndexEnd = p2pSubSVIndexBegin + nSubSVsP2P;
+        for (size_t p2pSubSVIndex = p2pSubSVIndexBegin;
+             p2pSubSVIndex < p2pSubSVIndexEnd; p2pSubSVIndex++) {
+            if (static_cast<size_t>(mpi_manager.getRank()) == p2pSubSVIndex)
+                continue; // don't need local sub state vector pointer
+            void *d_subSVP2P = nullptr;
+            const auto &dstMemHandle = ipcMemHandles[p2pSubSVIndex];
+            PL_CUDA_IS_SUCCESS(cudaIpcOpenMemHandle(
+                &d_subSVP2P, dstMemHandle, cudaIpcMemLazyEnablePeerAccess));
+            d_subSVsP2P.push_back(d_subSVP2P);
+            cudaEvent_t eventP2P = nullptr;
+            PL_CUDA_IS_SUCCESS(cudaIpcOpenEventHandle(
+                &eventP2P, ipcEventHandles[p2pSubSVIndex]));
+            remoteEvents.push_back(eventP2P);
+            subSVIndicesP2P.push_back(p2pSubSVIndex);
+        }
+
+        // set p2p sub state vectors
+        PL_CUSTATEVEC_IS_SUCCESS(custatevecSVSwapWorkerSetSubSVsP2P(
+            /* custatevecHandle_t */ handle,
+            /* custatevecSVSwapWorkerDescriptor_t */ svSegSwapWorker,
+            /* void** */ d_subSVsP2P.data(),
+            /* const int32_t* */ subSVIndicesP2P.data(),
+            /* cudaEvent_t */ remoteEvents.data(),
+            /* const uint32_t */ static_cast<int>(d_subSVsP2P.size())));
```

Comment:
Should this remain a `uint32_t` or be a `size_t`?
```suggestion
            /* const uint32_t */ static_cast<uint32_t>(d_subSVsP2P.size())));
```